newdle/api.py

```diff
@@ -77,6 +82,45 @@ def require_token():
     g.user = user
 
 
+@api.route('/avatar/<payload>')
+@allow_anonymous
+def user_avatar(payload):
+    user_info = avatar_info_from_payload(payload)
+    size = request.args.get('size')
+    email_hex = hashlib.md5(user_info['email'].encode()).hexdigest()
+    # make gravatar return 404 HTTP code instead of a default image
+    query_args = {'d': '404'}
+    if size is not None:
+        query_args['s'] = size
+
+    gravatar_url = f'https://gravatar.com/avatar/{email_hex}?{url_encode(query_args)}'
+    request_headers = {
+        header: value
+        for header, value in request.headers.items()
+        if header in ['if-modified-since']
+    }
+    resp = requests.get(gravatar_url, headers=request_headers)
+    if resp.status_code == 404:
+        return render_user_avatar(user_info['initial'], size)
```

Comment:
We can add some strong caching headers here, since the fallback never changes

Reply:
we could also just change the `send_file` cache_timeout from the default value to something more appropriate

Reply:
ah, I wasn't aware this is using `send_file`. then I think it may already add cache-related headers..

Reply:
it only adds cache-control ;)

Reply:
caching seems to work fine in the end, so I don't think we need to change anything :)